### PR TITLE
add GPG VERIFY true to production environment; add tests and fixed en…

### PR DESCRIPTION
**Pull request by acosferreira**
…v validation

# Description
Add a environment validation to set the GPG_VERIFY as true/false and avoid security issues on updates
# NEED TO WAIT FOR NEW TEMPLATE SIGNATURE

FIXES: 

## Type of change

What is it?

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

**Comment by acosferreira** [comment]
@rexwhite can you help me to resign the template after those changes?

File: pkg/services/updates.go [hunk 1]
```diff
@@ -262,7 +262,11 @@ func (s *UpdateService) CreateUpdate(id uint) (*models.UpdateTransaction, error)
 	remoteInfo.RemoteName = "rhel-edge"
 	remoteInfo.ContentURL = update.Repo.URL
 	remoteInfo.UpdateTransactionID = update.ID
-	remoteInfo.GpgVerify = "false"
+	if os.Getenv("SOURCES_ENV") == "prod" {
```
Comment (ldjebran) [inline pkg/services/updates.go:265]: Maybe this should be a configuration option, what do you think ?

File: pkg/services/updates_test.go [hunk 2]
```diff
@@ -673,12 +674,57 @@ var _ = Describe("UpdateService Basic functions", func() {
 		})
 
 		Context("when upload works", func() {
+			BeforeEach(func() {
+				err := os.Setenv("SOURCES_ENV", "true")
```
Comment (ldjebran) [inline pkg/services/updates_test.go:678]: I cannot get it, we are setting env to "true" but the condition in the function must be "prod" ?

File: pkg/services/updates_test.go [hunk 3]
```diff
@@ -673,12 +674,57 @@ var _ = Describe("UpdateService Basic functions", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				os.Unsetenv("SOURCES_ENV")
+			})
+			It("to build the template for PROD rebase properly", func() {
```
Comment (omaciel) [inline pkg/services/updates_test.go:685]: What I'm missing is a check to ensure that the template has the correct value for gpgcheck... what have I missed? :/

**Review by omaciel — COMMENTED** [review]
I have questions :/

File: pkg/services/updates_test.go (hunk 3 above)
Comment (de1987) [reply pkg/services/updates_test.go:685]: I was thinking about it too

sadly all of them require changing the main code

Comment (de1987) [reply pkg/services/updates_test.go:685]: maybe the easier way would be to create `TemplateService` that will allow to mock and spy on it

**Comment by rexwhite** [comment]
> @rexwhite can you help me to resign the template after those changes?

@acosferreira Sure thing - just let me know when you're ready!

File: templates/template_playbook_dispatcher_ostree_rebase_payload_prod.test.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+# This playbook is meant to be an all-in-one
```
[the remark this answers is not in the document]
Comment (acosferreira) [reply templates/template_playbook_dispatcher_ostree_rebase_payload_prod.test.yml:1]: this template is the expected result to assure that we are writing the correct value based on prod environment.
given the values we pass to the writeTemplate, the test is expecting have the same value of this template.
The others templates are covering no prod env

File: pkg/services/updates_test.go (hunk 3 above)
Comment (acosferreira) [reply pkg/services/updates_test.go:682]: this check is being doing inside of the template, based on the value that we set before the test, im expecting have the GpgVerify: "true" written on the template.

Comment (acosferreira) [reply pkg/services/updates_test.go:682]: I can also add new checks ;)

File: templates/template_playbook_dispatcher_ostree_rebase_payload_prod.test.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (acosferreira) [reply templates/template_playbook_dispatcher_ostree_rebase_payload_prod.test.yml:1]: where is being used
https://github.com/RedHatInsights/edge-api/pull/1894/files#diff-d8d295943afdc3f7001a2eb5b111aa69467ed482c360b4340904729afb157982R704

File: pkg/services/updates_test.go (hunk 3 above)
Comment (de1987) [reply pkg/services/updates_test.go:683]: Thanks, @acosferreira

File: templates/template_playbook_dispatcher_ostree_upgrade_payload.yml [hunk 5]
```diff
@@ -9,7 +9,7 @@
     ostree_remote_name: "@@ .GoTemplateRemoteName @@"
     ostree_changes_refs: "@@ .RemoteOstreeUpdate @@"
     os_tree_ref: "@@ .OSTreeRef @@"
-    ostree_gpg_verify: "false"
+    ostree_gpg_verify: "@@ .GoTemplateGpgVerify @@"
```
Comment (ldjebran) [inline templates/template_playbook_dispatcher_ostree_upgrade_payload.yml:12]: for this to work with signature, you need to add "/vars/ostree_gpg_verify"  to "insights_signature_exclude" section.

File: pkg/services/updates_test.go [hunk 6]
```diff
@@ -673,12 +674,55 @@ var _ = Describe("UpdateService Basic functions", func() {
 		})
 
 		Context("when upload works", func() {
+			var cfg *config.EdgeConfig
+			BeforeEach(func() {
+				os.Setenv("SOURCES_ENV", "prod")
+				cfg, _ = config.CreateEdgeAPIConfig()
```
Comment (ldjebran) [inline pkg/services/updates_test.go:680]: setting the env variable here with the current tests , has no effect, the config here is not the global config,

we can set TemplateRemoteInfo { GpgVerify: true } as the second test is setting false.

File: pkg/services/updates.go [hunk 7]
```diff
@@ -262,7 +262,8 @@ func (s *UpdateService) CreateUpdate(id uint) (*models.UpdateTransaction, error)
 	remoteInfo.RemoteName = "rhel-edge"
 	remoteInfo.ContentURL = update.Repo.URL
 	remoteInfo.UpdateTransactionID = update.ID
-	remoteInfo.GpgVerify = "false"
+	remoteInfo.GpgVerify = config.Get().GpgVerify
```
Comment (ldjebran) [inline pkg/services/updates.go:265]: We may still need to test that when calling writeTemplate from CreateUpdate we are passing the right gpgVerify value

Comment (ldjebran) [reply pkg/services/updates.go:265]: I propose to create a new function 
```go
func NewTemplateRemoteInfo(update *models.UpdateTransaction) TemplateRemoteInfo{
        ...
	return TemplateRemoteInfo{
		RemoteURL: update.Repo.URL,
		RemoteName: "rhel-edge",
		ContentURL: update.Repo.URL,
		UpdateTransactionID: update.ID,
		GpgVerify: config.Get().GpgVerify,
		OSTreeRef: update.Commit.OSTreeRef,
		RemoteOstreeUpdate: fmt.Sprint(update.ChangesRefs),
	}
}
```

use this function in CreateUpdate
```go
func (s *UpdateService) CreateUpdate(id uint) (*models.UpdateTransaction, error) {
     ....
     remoteInfo := NewTemplateRemoteInfo(update)
     ....
}
```

create a unittest for function NewTemplateRemoteInfo

what do you think ?

File: templates/template_playbook_dispatcher_ostree_upgrade_payload.yml (hunk 5 above)
Comment (acosferreira) [reply templates/template_playbook_dispatcher_ostree_upgrade_payload.yml:12]: thanks for the catch!

File: pkg/services/updates_test.go (hunk 6 above)
Comment (acosferreira) [reply pkg/services/updates_test.go:680]: With the current configuration we are setting the proper values and the gpg is being set to true and expected as result and covering both scenarios, true and false.

**Review by de1987 — DISMISSED** [review]

**Review by de1987 — APPROVED** [review]

**Review by ldjebran — DISMISSED** [review]

File: pkg/services/updates_test.go [hunk 8]
```diff
@@ -1580,4 +1627,30 @@ var _ = Describe("UpdateService Basic functions", func() {
...
+
+			orgID := faker.UUIDHyphenated()
+			update = &models.UpdateTransaction{
+				DispatchRecords: []models.DispatchRecord{},
+				OrgID:           orgID,
+				Commit:          &models.Commit{OSTreeRef: "ref"},
+				Repo:            &models.Repo{URL: "http://rh.com"},
+			}
+
+		})
```
Comment (ldjebran) [inline pkg/services/updates_test.go:1643]:
```go
AfterEach(func() {
    config.Get().GpgVerify = "false"
}) 
```

**Comment by acosferreira** [comment]
@rexwhite  We are ready to update the signature. Thanks!
As soon you send I'll move the pr from draft

**Review by ldjebran — APPROVED** [review]

**Pull request merged**